ccm_web/client/src/pages/FormatThirdPartyGradebook.tsx

```diff
@@ -0,0 +1,446 @@
+import SuccessCard from '../components/SuccessCard'
+import ThirdPartyGradebookConfirmationTable from '../components/ThirdPartyGradebookConfirmationTable'
+import WarningAlert from '../components/WarningAlert'
+import usePromise from '../hooks/usePromise'
+import { CanvasCourseSection, injectCourseName } from '../models/canvas'
+import { InvalidationType } from '../models/models'
+import { CCMComponentProps } from '../models/FeatureUIData'
+import CSVSchemaValidator, { SchemaInvalidation } from '../utils/CSVSchemaValidator'
+import FileParserWrapper, { CSVRecord } from '../utils/FileParserWrapper'
+import GradebookProcessor, {
+  GradebookInvalidation, GradebookUploadRecord, isGradebookUploadRecord, REQUIRED_LOGIN_ID_HEADER,
+  REQUIRED_ORDERED_HEADERS
+} from '../utils/GradebookProcessor'
+import { createOutputFileName } from '../utils/fileUtils'
+
+const useStyles = makeStyles((theme) => ({
+  root: {
+    padding: 25,
+    textAlign: 'left'
+  },
+  buttonGroup: {
+    marginTop: theme.spacing(1)
+  },
+  backButton: {
+    marginRight: theme.spacing(1)
+  },
+  stepper: {
+    textAlign: 'center',
+    paddingTop: '20px'
+  },
+  table: {
+    paddingLeft: 10,
+    paddingRight: 10
+  },
+  selectContainer: {
+    position: 'relative',
+    zIndex: 0,
+    textAlign: 'center'
+  },
+  uploadContainer: {
+    textAlign: 'center'
+  },
+  reviewContainer: {
+    textAlign: 'center'
+  },
+  backdrop: {
+    zIndex: theme.zIndex.drawer + 1,
+    color: '#FFF',
+    position: 'absolute'
+  }
+}))
+
+enum FormatGradebookStep {
+  Select = 0,
+  Upload = 1,
+  Review = 2,
+  Confirmation = 3
+}
+
+interface FormatThirdPartyGradebookProps extends CCMComponentProps {}
+
+export default function FormatThirdPartyGradebook (props: FormatThirdPartyGradebookProps): JSX.Element {
+  const classes = useStyles()
+
+  const csvParser = new FileParserWrapper(
+    Object.assign(FileParserWrapper.defaultParseConfigOptions, { transformHeader: undefined })
+  )
+
+  const [activeStep, setActiveStep] = useState<FormatGradebookStep>(FormatGradebookStep.Select)
+  const [sections, setSections] = useState<SelectableCanvasCourseSection[] | undefined>(undefined)
+  const [selectedSections, setSelectedSections] = useState<SelectableCanvasCourseSection[] | undefined>(undefined)
+  const [studentLoginIds, setStudentLoginIds] = useState<string[] | undefined>(undefined)
+  const [file, setFile] = useState<File | undefined>(undefined)
+  const [records, setRecords] = useState<GradebookUploadRecord[] | undefined>(undefined)
+  const [processedRecords, setProcessedRecords] = useState<GradebookUploadRecord[] | undefined>(undefined)
+  const [assignmentHeader, setAssignmentHeader] = useState<string | undefined>(undefined)
+
+  const [schemaInvalidations, setSchemaInvalidations] = useState<SchemaInvalidation[] | undefined>(undefined)
+  const [gradebookInvalidations, setGradebookInvalidations] = useState<GradebookInvalidation[] | undefined>(undefined)
```

Comment:
Observation: lot of `undefined` is used as default value, why is this a better option than empty string or list ( just as example)

Reply:
This is a habit of mine. I'm not sure if it's strictly better, but it feels more intuitive to me to be able to differentiate between `undefined` or something defined. If you use an empty string or array, you don't actually know that it's been updated (since the result of some process could be an empty value, like a search that returns an empty array). You could use `null` instead, but `undefined` felt more like an initial or cleared state.

Reply:
One result of my approach is that I can do checks like `value !== undefined`, whereas if I used an empty array, I'd have to do `value.length > 0` and I still wouldn't be able to tell if it had been updated or not.

Reply:
OK. I am convinced.